src/lando/pushlog/pushlog.py

```diff
@@ -0,0 +1,123 @@
+import logging
+from contextlib import contextmanager
+from typing import Optional
+
+from django.db import transaction
+
+from lando.main.models.repo import Repo
+from lando.main.scm.commit import Commit as SCMCommit
+from lando.pushlog.models.commit import Commit
+from lando.pushlog.models.push import Push
+
+logger = logging.getLogger(__name__)
+
+
+@contextmanager
+def PushLogForRepo(repo: Repo, user: str):
+    """
+    Context manager allowing to incrementally build push information, and only submit it
+    when complete.
+
+    WARNING: Do not use record_push() on the returned PushLog, as the context manager
+    will take care of it automatically. Calling it separately may lead to duplicates.
+    """
+    pushlog = PushLog(repo, user)
+    try:
+        yield pushlog
+    except Exception as e:
+        logger.error(f"Push aborted for {user}: {e}")
+        raise (e)
+    else:
+        # Only record the whole push on success.
+        try:
+            pushlog.record_push()
+        except Exception as e:
+            # We keep a record of the Pushlog in the extra, in addition to printing
+            # details in the log.
+            logger.error(
+                f"Failed to record push log due to: {e}\n{pushlog}",
+                extra={"pushlog": pushlog},
+            )
+            raise e
+
+
+class PushLog:
+    repo: Repo
+    push: Push
+    user: str
+
+    confirmed: bool = False
+
+    commits: list
+
+    def __init__(
+        self,
+        repo: Repo,
+        user: str,
+        commits: list = [],
```

Comment:
There's a Python gotcha here, should not specify a mutable argument ([]), which you hint at below.

```suggestion
        commits: list = None,
```

Reply:
Ah! Yes. I handle the problem just below, by creating a fresh new list https://github.com/mozilla-conduit/lando/pull/196/files/4a97d5c23c741a958303178ff3252d37838b744d#diff-b518a05517dedc31aceb83ab380e3bd35de03582d5fc1fa528c716e639882463R62-R65

I'm reluctant to use `None` in the prototype, as this would require to make it an `Optional[list]`, unless you insist?

Reply:
It's very conventional to not set a `[]` and instead set `None`. I think it's also better not to add a workaround to satisfy type hint annotations.

Lastly, passing an empty list here would also trigger lines 66-69, so I think it is better all around to set it to `None` and always set the default within the function.